torch/_inductor/fx_passes/group_batch_fusion.py

```diff
@@ -120,19 +124,98 @@ def fuse(self, graph, subset):
             graph.erase_node(original_mm)
 
 
-class BatchLayernormFusion(BatchFusion):
+class BatchLinearLHSFusion(BatchFusion):
     """
-    Batch layer norm fusion in pre grad pass
+    Batch linear left-hand side fusion. This pass tries to fuse the following patterns:
+
+        torch.nn.functional.linear(x, w1), linear(x, w2),... * linear(x, wn)
+        -> torch.mm(x, torch.cat([w1, w2,... * wn]).transpose(0, 1))
+
+    We have a separate pass to eliminate contiguous transpose in a generic way.
     """
 
-    def is_node_meta_valid(self, node):
-        if node is None:
-            return True
+    def _linear_node_can_be_fused(self, node):
+        input = get_arg_value(node, 0, "input")
+        weight = get_arg_value(node, 1, "weight")
+        return (
+            is_node_meta_valid(node)
+            and len(input.meta["example_value"].shape) == 2
+            and len(weight.meta["example_value"].shape) == 2
+        )
+
+    def match(self, node):
+        if CallFunctionVarArgs(torch.nn.functional.linear).match(
+            node
+        ) and self._linear_node_can_be_fused(node):
+            input = get_arg_value(node, 0, "input")
+            bias = get_arg_value(node, 2, "bias")
+            group_key = ("batch_linear_lhs", bias is None, input)
         else:
-            if "example_value" in node.meta:
-                return True
+            group_key = None
+        return group_key
+
+    def fuse(self, graph, subset):
+        batch_nodes = []
+        batch_input = None
+        batch_weights = []
+        batch_biases = []
+        split_sections = []
+        for node in subset:
+            input = get_arg_value(node, 0, "input")
+            weight = get_arg_value(node, 1, "weight")
+            bias = get_arg_value(node, 2, "bias")
+            batch_nodes.append(node)
+            if batch_input is None:
+                batch_input = input
             else:
-                return False
+                assert batch_input is input
+            batch_weights.append(weight)
+            if bias:
+                batch_biases.append(bias)
+            split_sections.append(weight.meta["example_value"].shape[0])
```

Comment:
Noob question. Somehow I still saw this warning in my local run. "[2023-08-02 20:11:05,605] torch._inductor.fx_passes.split_cat: [WARNING] example value absent for node: split_15"

Reply:
I saw it as well, but it's from the split cat pass. Some nodes don't have example value and we won't apply split-cat patterns. We should look at why the example value absent in the input graph, which should be a dynamo issue IMO. I'll follow up this in a separate thread.